adds windows modding tool

Doronimmo opened the pull request.

evanbowman commented:
Looks pretty good! I'm going to take a look at the scripts again and see if I can make things even easier. Ideally, people who want to build the game would only need to run a single python script, let me see...

evanbowman commented:
In general, here's what I'm thinking:
- I think we should rename the `join fs_engine.py` script to `join_fs_engine.py`. And, if join_fs_engine.py works fine in linux/Mac OS, then I think that this script should be used by everyone who wants to edit the scripts, rather than just windows users. Having everyone follow the same steps would simplify the instructions.
- I'd like to add an additional script, maybe just called `build.py`, that runs the other two scripts, then we can have people just run a single build script to rebuild Skyland.gba.

I'm going to check out your branch and play with the scripts a bit to make sure it works the same on Mac and linux

Doronimmo commented:
> * `join fs_engine.py` script to `join_fs_engine.py`

it is already named that way

evanbowman commented:
> > * `join fs_engine.py` script to `join_fs_engine.py`
> 
> it is already named that way

In the set of changes in GitHub, it shows up with a space rather than an underscore for me: https://github.com/evanbowman/skyland-beta/pull/55/files

How does it look for you?

evanbowman commented:
Screenshot:
<img width="375" alt="Screenshot 2024-11-01 at 1 09 21 PM" src="https://github.com/user-attachments/assets/9abfed09-49eb-4b94-b36e-5ec4d5af24a3">

evanbowman commented:
> > * `join fs_engine.py` script to `join_fs_engine.py`
> 
> it is already named that way edit: did not see the underscore

No problem!

Doronimmo commented:
renamed the join_fs_engine.py

Doronimmo commented:
currently not editing the readme.md in favor of waiting for the build.py

evanbowman commented:
I think if you add this to tools/build.py, it should be enough:
```python
import encode_files
import join_fs_bin
```
Then people will only need to run one build script to rebuild the game

Doronimmo commented:
from where to run the encode_files.py?
if I run it from tools dir, it will will make a file, but the resulting ROM will be a white screen

evanbowman commented:
ah, so encode_files should be run from the upper directory, like:
```
python3 tools/encode_files.py
```

evanbowman commented:
There are a couple of places in the game's documentation besides the readme where I refer to the `cat` command, we should update documentation in a few different places, let me see...

Doronimmo commented:
I run the encode_files from the tools dir it will make an fs.bin in source/tools/fs.bin
but the join_fs_engine looks for it in the source/fs.bin

evanbowman commented:
https://github.com/evanbowman/skyland-beta/blob/71d8ec0ba4bd556da3de7af3a23d7f60178cac97/tools/unpack_rom.py#L11

And then, there's another readme in the game's resource bundle, in addition to the README.md created for the GitHub page:
https://github.com/evanbowman/skyland-beta/blob/6e586354ae9c4ddf425658b814f738420d0023b3/readme.txt#L20

evanbowman commented:
> I run the encode_files from the tools dir it will make an fs.bin in source/tools/fs.bin but the join_fs_engine looks for it in the source/fs.bin

Maybe I should change encode_files to produce an error if run from the tools directory.

Doronimmo commented:
the skyland.gba results in a white screen for me, it seems to be the same file as the engine

Doronimmo commented:
I dont know what has changed, it worked before

evanbowman commented:
> I dont know what has changed, it worked before

Yeah, that's very strange. So you're in the top level directory, and running tools/encode_files.py? I'm unsure what to suggest as I can't see your files. If you zip everything up I can take a look

Doronimmo commented:
[skyland.zip](https://github.com/user-attachments/files/17602646/skyland.zip)

Doronimmo commented:
windows users can use the cat command with wsl

Doronimmo commented:
> windows users can use the cat command with wsl

but it takes a while to install and requires admin previliges

evanbowman commented:
```
$ ~ % cd Downloads/skyland 
$ skyland % ls
Skyland.gba		SkylandEngine.gba	encode_files.py		fs.extracted.bin	join_fs_engine.py	lisp_symtab.dat		scripts			tools
Skyland.sav		boot.ini		fs.bin			help			licenses		readme.txt		strings			unpack_rom.py
$ skyland % python3 tools/encode_files.py 
creating fs.bin...
encoding 357 files...
encoded 18881676 bytes
done!
$ skyland % mv join_fs_engine.py tools 
$ skyland % python3 tools/join_fs_engine.py 
$ skyland % ls
Skyland.gba		SkylandEngine.gba	encode_files.py		fs.extracted.bin	licenses		readme.txt		strings			unpack_rom.py
Skyland.sav		boot.ini		fs.bin			help			lisp_symtab.dat		scripts			tools
```

I ran encode_files.py from the top level directory, worked fine. Then I moved join_fs_engine.py into tools/, ran it, and it worked fine.

Doronimmo commented:
> ```
> $ ~ % cd Downloads/skyland 
> $ skyland % ls
> Skyland.gba		SkylandEngine.gba	encode_files.py		fs.extracted.bin	join_fs_engine.py	lisp_symtab.dat		scripts			tools
> Skyland.sav		boot.ini		fs.bin			help			licenses		readme.txt		strings			unpack_rom.py
> $ skyland % python3 tools/encode_files.py 
> creating fs.bin...
> encoding 357 files...
> encoded 18881676 bytes
> done!
> $ skyland % mv join_fs_engine.py tools 
> $ skyland % python3 tools/join_fs_engine.py 
> $ skyland % ls
> Skyland.gba		SkylandEngine.gba	encode_files.py		fs.extracted.bin	licenses		readme.txt		strings			unpack_rom.py
> Skyland.sav		boot.ini		fs.bin			help			lisp_symtab.dat		scripts			tools
> ```
> 
> I ran encode_files.py from the top level directory, worked fine. Then I moved join_fs_engine.py into tools/, ran it, and it worked fine.

OH I ran join_fs_engine.py from the /skyland dir all along

evanbowman commented:
> I think if create a file called tools/build.py, containing this code, it should be enough:
> 
> ```python
> import encode_files
> import join_fs_bin
> ```
> 
> Then people will only need to run one build script to rebuild the game

Can you make a build.py file in tools/, then run it from the top level directory and tell me if it works?

evanbowman commented:
It works for me, but I'd like to check if it works in windows too

Doronimmo commented:
can confirm it works!

evanbowman commented:
Looks good!

evanbowman commented on README.md:
```diff
@@ -58,7 +58,7 @@ Skyland is jam-packed with stuff, some of the highlights:
 # Modding (no C++ toolchain required!)
 
-I designed the Skyland to be moddable, and wrote all level scenarios in a custom scripting language called Skyland LISP. The script in modding/unpack_rom.py extracts the game's filesystem from the Skyland.gba ROM, producing a SkylandEngine.gba ROM and directories full of lisp scripts. The encode_files.py script in the extracted tools/ directory repacks the filesystem into a file called fs.bin. Th ...
+I designed the Skyland to be moddable, and wrote all level scenarios in a custom scripting language called Skyland LISP. The script in modding/unpack_rom.py extracts the game's filesystem from the Skyland.gba ROM, producing a SkylandEngine.gba ROM and directories full of lisp scripts. The encode_files.py script in the extracted tools/ directory repacks the filesystem into a file called fs.bin. Th ...
```
Can you make one small change?
```
 The encode_files.py script in the extracted tools/ directory repacks the filesystem into a file called fs.bin. Then, for linux simply run `cat SkylandEngine.gba fs.bin > Skyland.gba` to repackage the gba game and for windows run the join fs_engine.py in the tools directory. Skyland uses unicode internally, and you'll find much of the menu text in strings/strings.txt. You'll find the rest of the text scattered throughout the lisp scripts. See help/localization.txt for more info.
```
Let's delete this text and just tell people to run tools/build.py from the same directory where they run unpack_rom.py

Doronimmo replied:
👍

evanbowman replied:
Ah! I think I didn't explain correctly what I wanted changed in the readme. Actually, I'll merge your changes first and then edit the readme a bit. Thanks for your help with the python scripts!

Doronimmo replied:
No prob! happy to help!
